File: src/modules/ui/notification.cpp
/**
 * Copyright (c) 2012 - 2014 TideSDK contributors
 * http://www.tidesdk.org
 * Includes modified sources under the Apache 2 License
 * Copyright (c) 2008 - 2012 Appcelerator Inc
 * Refer to LICENSE for details of distribution and use.
 **/

#include "notification.h"

namespace ti
{

Notification::Notification() : AccessorObject("Notification"),
    timeout(-1)
{
    SetMethod("setTitle", &Notification::_SetTitle);
    SetMethod("setMessage", &Notification::_SetMessage);
    SetMethod("setIcon", &Notification::_SetIcon);
    SetMethod("setTimeout", &Notification::_SetTimeout);
    SetMethod("setDelay", &Notification::_SetTimeout);
    SetMethod("show", &Notification::_Show);
    SetMethod("hide", &Notification::_Hide);

    SetBool("nativeNotifications", Notification::InitializeImpl());
    this->CreateImpl();
}

Notification::~Notification()
{
    this->DestroyImpl();
}

void Notification::_SetTitle(const ValueList& args, ValueRef result)
{
    args.VerifyException("setTitle", "s");
    this->title = args.GetString(0);
}

void Notification::_SetMessage(const ValueList& args, ValueRef result)
{
    args.VerifyException("setMessage", "s");
    this->message = args.GetString(0);
}

void Notification::_SetIcon(const ValueList& args, ValueRef result)
{
    args.VerifyException("setIcon", "s");
    this->iconURL = args.GetString(0);
}

void Notification::_SetTimeout(const ValueList& args, ValueRef result)
{
    args.VerifyException("setTimeout", "i");
    this->timeout = args.GetInt(0);
}

void Notification::_Show(const ValueList& args, ValueRef result)
{
    result->SetBool(this->ShowImpl());
}

void Notification::_Hide(const ValueList& args, ValueRef result)
{
    result->SetBool(this->HideImpl());
}

void Notification::Configure(TiObjectRef properties)
{
    this->title = properties->GetString("title");
    this->message = properties->GetString("message");
    this->iconURL = properties->GetString("icon");
    this->timeout = properties->GetInt("timeout", -1);
}

}

File: tool/drillbit/Resources/tests/notification/notification.js
/**
 * Copyright (c) 2012 - 2014 TideSDK contributors 
 * http://www.tidesdk.org
 * Includes modified sources under the Apache 2 License
 * Copyright (c) 2008 - 2012 Appcelerator Inc
 * Refer to LICENSE for details of distribution and use.
 **/

describe("Notification tests", {

  test_notification_object: function () {
    var wnd = Ti.UI.createNotification(parent.window);

    value_of(wnd.hide).should_be_function();
    value_of(wnd.setDelay).should_be_function();
    value_of(wnd.setIcon).should_be_function();
    value_of(wnd.setMessage).should_be_function();
    value_of(wnd.setTitle).should_be_function();
    value_of(wnd.show).should_be_function();
    value_of(wnd).should_be_object();
  },

  test_it_with_old_icon_as_async: function (callback) {
    value_of(Ti.UI.createNotification).should_be_function();

    // create a notification object
    var parent = Ti.UI.getCurrentWindow();
    var wnd = Ti.UI.createNotification(parent.window);
    wnd.setTitle("title");
    wnd.setMessage("this is a message with the old drillbit icon");
    wnd.setIcon("app://logo_large.png");
    wnd.setDelay(5000);


    Ti.API.debug("attempting to show the notification");
    wnd.show();

    timer = setTimeout(function () {
      try {
        Ti.API.debug("hiding the notification");
        wnd.hide();
        callback.passed();
      } catch (e) {
        Ti.API.debug("hiding the notification failed with an exception");
        callback.failed();
      }
    }, 1000);
  },

  test_it_with_new_icon_as_async: function (callback) {
    value_of(Ti.UI.createNotification).should_be_function();

    // create a notification object
    var parent = Ti.UI.getCurrentWindow();
    var wnd = Ti.UI.createNotification(parent.window);
    wnd.setTitle("title");
    wnd.setMessage("this is a message with the new tidesdk icon");
    wnd.setIcon("app://logo_small.png");
    wnd.setDelay(5000);


    Ti.API.debug("attempting to show the notification");
    wnd.show();

    timer = setTimeout(function () {
      try {
        Ti.API.debug("hiding the notification");
        wnd.hide();
        callback.passed();
      } catch (e) {
        Ti.API.debug("hiding the notification failed with an exception");
        callback.failed();
      }
    }, 1000);
  },

  test_non_utf8_as_async: function (callback) {
    value_of(Ti.UI.createNotification).should_be_function();

    // create a notification object
    var parent = Ti.UI.getCurrentWindow();
    var wnd = Ti.UI.createNotification(parent.window);
    wnd.setTitle("title");
    wnd.setMessage("Стоял он, дум великих полн");
    wnd.setIcon("app://logo_small.png");
    wnd.setDelay(5000);


    Ti.API.debug("attempting to show the notification");
    wnd.show();

    timer = setTimeout(function () {
      try {
        Ti.API.debug("hiding the notification");
        wnd.hide();
        callback.passed();
      } catch (e) {
        Ti.API.debug("hiding the notification failed with an exception");
        callback.failed();
      }
    }, 1000);
  }
});
